include/deal.II/matrix_free/operators.h

```diff
@@ -328,6 +328,101 @@ namespace MatrixFreeOperators
 
 
 
+  /**
+   * This class implements the operation of the action of a Laplace matrix.
+   *
+   * @author Denis Davydov, 2016
+   */
+  template <int dim, int fe_degree, int n_q_points_1d = fe_degree+1, int n_components = 1, typename Number = double>
+  class LaplaceOperator : public Base<dim, Number>
+  {
+  public:
+
+    /**
+     * Constructor.
+     */
+    LaplaceOperator ();
+
+    /**
+     * The diagonal is approximated by computing a local diagonal matrix per element
+     * and distributing it to the global diagonal. This will lead to wrong results
+     * on element with hanging nodes but is still an acceptable approximation
+     * to be used in preconditioners.
+     */
+    virtual void compute_diagonal ();
+
+    /**
+     * Set heterogeneous scalar coefficient @p scalar_coefficient to be used at
+     * the quadrature points. The Table should be of correct size, consistent
+     * with the total number of quadrature points in <code>dim</code>-dimensions,
+     * controlled by @p n_q_points_1d template parameter. Here,
+     * <code>(*scalar_coefficient)(cell,q)</code> corresponds to the value of the
+     * coefficient, where <code>cell</code> is the cell number and <code>q</code>
+     * is the number of quadrature point.
+     *
+     * Such tables can be initialized by
+     * <code>
+     * std::shared_ptr<Table<2, VectorizedArray<double> > > coefficient;
+     * coefficient = std::make_shared<Table<2, VectorizedArray<double> > >();
+     * {
+     *   FEEvaluation<dim,fe_degree,n_q_points_1d,1,double> fe_eval(mf_data);
+     *   const unsigned int n_cells = mf_data.n_macro_cells();
+     *   const unsigned int n_q_points = fe_eval.n_q_points;
+     *   coefficient->reinit(n_cells, n_q_points);
+     *   for (unsigned int cell=0; cell<n_cells; ++cell)
+     *     {
+     *       fe_eval.reinit(cell);
+     *       for (unsigned int q=0; q<n_q_points; ++q)
+     *         (*coefficient)(cell,q) = function.value(fe_eval.quadrature_point(q));
+     *     }
+     * }
+     * </code>
+     * where `<code>mf_data</code> is MatrixFree object and <code>function</code>
+     * is a function which provides the following method
+     * <code>VectorizedArray<double> value(const Point<dim, VectorizedArray<double> > &p_vec)</code>.
+     *
+     * If this function is not called, coefficient is assumed to be unity.
+     */
+    void set_coefficient(std_cxx11::shared_ptr<Table<2, VectorizedArray<Number> > > scalar_coefficient );
```

Comment:
It's ok to just pass by value here, but you can save one copy if you passed by const reference.